apiserver/internal/handlers/resources/download/downloader.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package download
+
+import (
+	"context"
+	"crypto/sha512"
+	"encoding/hex"
+	"io"
+	"os"
+
+	"github.com/juju/juju/core/logger"
+	"github.com/juju/juju/internal/errors"
+)
+
+const (
+	// ErrUnexpectedHash is returned when the hash of the downloaded resource
+	// does not match the expected hash.
+	ErrUnexpectedHash = errors.ConstError("downloaded resource has unexpected hash")
+
+	// ErrUnexpectedSize is returned when the size of the downloaded resources
+	// does not match the expected size.
+	ErrUnexpectedSize = errors.ConstError("downloaded resource has unexpected size")
+)
+
+// FileSystem defines a file system for modifying files on a users system.
+type FileSystem interface {
+	// CreateTemp creates a new temporary file in the directory dir,
+	// opens the file for reading and writing, and returns the resulting file.
+	CreateTemp(dir, pattern string) (*os.File, error)
+
+	// Open opens the named file for reading. If successful, methods on
+	// the returned file can be used for reading; the associated file
+	Open(name string) (*os.File, error)
+
+	// Remove removes the named file or (empty) directory.
+	Remove(name string) error
+}
+
+// Downloader validates a resource blob.
+type Downloader struct {
+	logger     logger.Logger
+	fileSystem FileSystem
+}
+
+// NewDownloader returns a new resource blob validator.
+func NewDownloader(logger logger.Logger, fileSystem FileSystem) *Downloader {
+	return &Downloader{
+		logger:     logger,
+		fileSystem: fileSystem,
+	}
+}
+
+// Download takes a request body ReadCloser containing a resource blob and
+// checks that the size and hash match the expected values. It downloads the
+// blob to a temporary file and returns a ReadCloser that deletes the
+// temporary file on closure.
```

Comment:
This feels like it can be used for resources, charms and agent binaries. Is there any reason it should be specific to resources?

Reply:
The hash is 384 for resources, whereas for those it is 256, other than that there is no major difference. These could maybe be merged in the future.